trieste/ask_tell_optimization.py

```diff
@@ -0,0 +1,181 @@
+# Copyright 2021 The Trieste Contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This module contains the Ask/Tell API for users of Trieste who would like to
+perform Bayesian Optimization with external control of the optimization loop.
+"""
+
+
+from __future__ import annotations
+
+from typing import Dict, Generic, Mapping, Optional, TypeVar, cast
+
+from .acquisition.rule import AcquisitionRule
+from .bayesian_optimizer import Record
+from .data import Dataset
+from .models import ModelSpec, create_model
+from .observer import OBJECTIVE
+from .space import SearchSpace
+from .types import TensorType
+from .utils import map_values
+
+S = TypeVar("S")
+""" Unbound type variable. """
+
+SP = TypeVar("SP", bound=SearchSpace)
+""" Type variable bound to :class:`SearchSpace`. """
+
+
+class AskTellOptimizer(Generic[S, SP]):
+    """
+    This class provides Ask/Tell optimization interface. It is designed for those use cases
+    when control of the optimization loop by Trieste is impossible.
+    That could happen because objective function cannot be implemented/called in Python, or if
+    optimization state cannot be maintained in memory.
+    For more details about the Bayesian Optimization routine, refer to :class:`BayesianOptimizer`.
+    """
+
+    def __init__(
+        self,
+        search_space: SP,
+        datasets: Mapping[str, Dataset] | Dataset,
+        model_specs: Mapping[str, ModelSpec] | ModelSpec,
+        fit_model: bool,
+        acquisition_rule: AcquisitionRule[S, SP],
+        acquisition_state: Optional[S] = None,
+    ):
+        """
+        :param search_space: The space over which to search.
+        :param datasets: The known observer query points and observations for each tag.
+        :param model_specs: The model to use for each :class:`~trieste.data.Dataset` in
+            ``datasets``.
+        :param fit_model: If `True`, models passed in will be optimized on the given data.
+            If `False`, the models are assumed to be optimized already.
+        :param acquisition_rule: The acquisition rule, which defines how to search for a new point
```

Comment:
do we want to default to `EfficientGlobalOptimization` as in `bayesian_optimizer`?

Reply:
Uri also suggested the same, makes sense to keep API similar. agreed!